device_doctor/lib/src/device.dart

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'config.dart';
+import 'ios_device.dart';
+
+/// The root of the API for controlling devices.
+DeviceDiscovery get devices => DeviceDiscovery();
+
+/// Operating system on the devices that this host is configured to test.
+enum DeviceOperatingSystem { android, ios }
+
+/// Discovers available devices and chooses one to work with.
+abstract class DeviceDiscovery {
+  factory DeviceDiscovery() {
+    switch (config.deviceOperatingSystem) {
+      case DeviceOperatingSystem.ios:
+        return IosDeviceDiscovery();
+      default:
+        throw StateError('Unsupported device operating system: {config.deviceOperatingSystem}');
+    }
+  }
+
+  /// Lists all available devices' IDs.
+  Future<List<Device>> discoverDevices({int retriesDelayMs = 10000});
+
+  /// Checks the health of the available devices.
+  Future<Map<String, HealthCheckResult>> checkDevices();
+
+  /// Recovers the device.
+  Future<void> recoverDevice();
+}
+
+/// A proxy for one specific device.
+abstract class Device {
```

Comment:
This API seems too specific to implement beyond phones. Would `initialize``recover`, and `collect_logs` make more sense? Some of these methods make more sense as part of a child interface like `PhoneDevice`.

Reply:
The device here is meant for phone device. Updated the doc to reflect this.

Reply:
My use case is how can we generalize this to also work with Fuchsia devices later on. With this approach, we'll need to further abstract this class when we do add that support.

Reply:
Talked offline, will leave the class as is now. May update/extend the class when we get to the point to support fuchsia devices.